### PR TITLE
Checkout: Switch Web Payments to Stripe

### PR DESCRIPTION
**Pull request by sirbrillig**
The current Web Payment system (eg: Apple Pay) has special logic for creating PaymentRequest objects, showing a payment form, and sending that data to the server for a special checkout flow. 

As part of replacing Paygate usage in Calypso, this PR swaps out most of the custom code for the Stripe Elements _Payment Request Button_. This button has its own Payment Request object that abstracts over Apple Pay, Google Pay, etc, and integrates tightly with our other Stripe Elements payment flow (added in #34469) to create a Stripe Payment Method. It then sends that token to the server in the same manner as the current credit card flow.

apple-pay-form

#### Testing instructions

1. Sandbox the store.
2. Use a proxy to forward requests from `https://wordpress.com` to localhost. This can be a complex step. Let me know if you have any questions about doing this. It's necessary because the button will only appear on a secure connection and on a verified url.
3. Make sure you have a real credit card added to Apple Pay in your local Safari. Look online for how to do this if you don't have it set up already or let me know and I can help.
3. Add a plan to your cart. Make sure you see the "dev" logo in the lower right of the screen.
4. Visit the checkout page in Safari. Apple Pay exists only in Safari on Mac OS.
5. Click the Apple Pay tab above the checkout form.
6. Click the Pay button.
7. When the Apple Pay sheet appears, authorize the payment. Don't worry, you won't be charged as long as you are sandboxing the store.
8. Verify that the transaction goes through correctly and that you are redirected to the next step of the checkout process (which may be a special offer).

### To do

- [x] Take care of updates to paymentRequest: https://github.com/stripe/react-stripe-elements/issues/284
- [x] Test errors and failures
- [x] Try to abstract away all the memoizations
- [x] Make sure country/postal are passed correctly
- [x] Verify purchase record (receipt, ownership, etc.) is correct on the server
- [x] Consider renewal flows
- [x] Convert amounts to correct values: D31850-code
- [x] Add better loading indicator for button
- [x] Disable button when country/postal are empty
- [x] Verify non-US currencies are shown and work

### CODE REVIEW

**Comment by sirbrillig** [comment]
I think debouncing broke a test. I'll figure that out shortly.

**Comment by sirbrillig** [comment]
This also works with Google Pay if the config flag is enabled!

Screen Shot 2019-08-30 at 2 55 09 PM

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 1]
```diff
@@ -28,523 +26,383 @@ import {
...
 	};
+	if ( paymentType === 'apple-pay' ) {
+		return <button className="web-payment-box__apple-pay-button" onClick={ onClick } />;
+	}
+	return (
+		<button
+			className="web-payment-box__web-pay-button button checkout__pay-button-button button is-primary button-pay pay-button__button"
+			onClick={ onClick }
+		>
+			{ translate( 'Select a payment card' ) }
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: Hi! I've found a possible matching string that has already been translated **22** times:
`translate( 'Select a payment card', { context: 'Loading state on /checkout'} )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/blocks/credit-card-form/helpers.js [hunk 2]
```diff
@@ -139,8 +139,14 @@ export function getParamsForApi( cardDetails, cardToken, stripeConfiguration, ex
 
 export function useDebounce( value, delay ) {
 	const [ debouncedValue, setDebouncedValue ] = useState( value );
-	const debounced = useRef( debounce( newValue => setDebouncedValue( newValue ), delay ) );
```
Comment (DavidRothstein) [inline client/blocks/credit-card-form/helpers.js]: It wasn't obvious to me why this needs to move away from the `lodash` version, but I'll trust there's a reason :)

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 3]
```diff
@@ -28,523 +26,383 @@ import {
...
+export function WebPaymentBox( {
+	cart,
+	countriesList,
+	onSubmit,
+	presaleChatAvailable,
+	translate,
+	disablePostalCodeDebounce,
+	children,
+} ) {
+	const paymentMethod = useMemo( () => detectWebPaymentMethod(), [] );
```
Comment (DavidRothstein) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: Is the `useMemo` necessary here?  `detectWebPaymentMethod` should be fast (and it takes care of caching internally).

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 4]
```diff
@@ -28,523 +26,383 @@ import {
...
+	currency,
+	shouldDisplayItems,
+} ) {
+	const displayItems = useMemo(
+		() => getDisplayItems( { translate, sub_total_integer, total_tax_integer } ),
+		[ translate, sub_total_integer, total_tax_integer ]
+	);
+	const paymentRequestOptions = useMemo(
+		() => ( {
+			country: countryCode || 'US',
```
Comment (DavidRothstein) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: I think this may be sending the wrong country.  If I'm reading the code right, it looks like it's passing in the billing country, but based on https://stripe.com/docs/stripe-js/reference#stripe-payment-request it looks like it should be the Stripe account country.

In the old version, we had code that contacted `wpcom.undocumented().paygateConfiguration()` to get the best guess of which Automattic Stripe account (and therefore country) the payment would go through.  It's not clear to me if the same code would work here or not (and I'm also not sure offhand how `stripe.js` payments work with different Stripe accounts, compared to Paygate ones?).  Nor am I sure how important it is to send the correct country here... although I remember hearing somewhere that it can potentially cause Apple Pay's bypass of Strong Customer Authentication to fail.

Bottom line, though - it probably shouldn't be the billing country.

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 5]
```diff
@@ -28,523 +26,383 @@ import {
...
-const debug = debugFactory( 'calypso:checkout:payment:apple-pay' );
+const debug = debugFactory( 'calypso:checkout:payment:web-payment-box' );
 
-/**
- * Supported card types.
- */
-const SUPPORTED_NETWORKS = [ 'visa', 'masterCard', 'discover', 'amex', 'jcb', 'chinaUnionPay' ];
-const APPLE_PAY_MERCHANT_IDENTIFIER = config( 'apple_pay_merchant_id' );
 const PAYMENT_REQUEST_OPTIONS = {
 	requestPayerName: true,
```
Comment (DavidRothstein) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: I don't understand why, but this is apparently forcing us to collect not just the name but also the entire billing address.

Stripe even documents that: https://stripe.com/docs/stripe-js/reference#stripe-payment-request ("We highly recommend you collect at least one of name, email, or phone as this also results in collection of billing address for Apple Pay")

With regular Apple Pay (our current implementation) it just collects the name, which is all we need.  I'm pretty sure it only requires the billing address if you directly ask for it.

This change in behavior is kind of bad because it means we're requiring Apple Pay users to type in lots of extra information that we don't even require from credit card users (although at least hopefully many of them have typed it in and saved it already)... maybe we need to reach out to Stripe to understand what is going on here.

Comment (DavidRothstein) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: And another side effect of that (I think) is that it results in the postal code that the user typed before opening Apple Pay being overwritten (in the data saved server-side) with the one that came from the billing address in Apple Pay -- in the case where they differ.

I think in general we have been trying to consistently save the one the user typed since that is the one used to calculate taxes on checkout.

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 6]
```diff
@@ -28,523 +26,383 @@ import {
...
-				text: translate( 'Sending your purchase', {
-					context: 'Loading state on /checkout',
-				} ),
-			};
-		};
-		const completingState = () => {
-			return {
-				default: false,
-				disabled: true,
-				text: translate( 'Completing your purchase', {
```
Comment (DavidRothstein) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: Is there a way to keep these button states in place (or alternatively replace them with some other kind of progress indicator)?

The experience seems pretty bad without them, since once you complete the Apple Pay payment it just closes the payment sheet and shows you the page underneath, which still has the "Buy with Apple Pay" button showing and no indicator that your payment is processing.  Especially for domain purchases, where it can take a while to process the payment, this makes it look like nothing is happening.

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 7]
```diff
@@ -28,523 +26,383 @@ import {
...
-		switch ( paymentMethod ) {
-			case WEB_PAYMENT_APPLE_PAY_METHOD:
-				if ( buttonState.default ) {
-					button = (
-						<button
-							type="submit"
-							onClick={ event =>
-								! countryCode || ! postalCode
-									? this.displaySubmissionError(
-											translate( 'Please specify a country and postal code.' ),
```
Comment (DavidRothstein) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: Is there a way to keep this?

The user experience without it (where the button is just grayed out with no indication of why it can't be submitted) definitely seems much worse.

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 8]
```diff
@@ -28,523 +26,383 @@ import {
...
+} ) {
+	const { card, id } = paymentMethod;
+
+	const newCardDetails = {
+		payment_key: id,
+		name: payerName,
+		country: countryCode,
+		'postal-code': postalCode,
+		card_brand: card.brand,
+		card_display_name: `${ card.brand } ${ card.last4 }`,
```
Comment (DavidRothstein) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: We probably don't need this -- I think the current server-side code that parses it only runs for Paygate anyway.  It's there now just because for raw Apple Pay purchases, the card `displayName` they provide is the only way to get the actual last 4 digits of the user's card.  But it looks like Stripe already takes care of returning that data in a cleaner format in their API.

File: client/blocks/credit-card-form/helpers.js (hunk 2 above)
Comment (sirbrillig) [reply client/blocks/credit-card-form/helpers.js]: It actually probably doesn't. I was trying to fix the tests and I thought that maybe the lodash debounce was causing problems, but it turns out it's just that enzyme shallow mounting will not run `useEffect` hooks yet. I could probably remove this change.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 3 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Oh! I just read that [it can be expensive](https://github.com/Automattic/wp-calypso/blob/e5bf30394355680532797aa31de89d5835886f56/client/lib/web-payment/index.js#L31) but I didn't actually notice that it was doing the caching there. I can remove this too.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 4 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: I remember when I was first refactoring this section that I wasn't sure if it should be the billing or the processor country, but for some reason I decided it must be the billing country. It seems strange that it would want the processor country, since the Stripe object should already have that in its state. That said, 

> The two-letter country code of your Stripe account (e.g., 'US').

seems pretty clear that you're right. Weird. Ok, I can change it.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 5 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: 🤔 Thanks. I need to take a deeper look at this. I thought we were sending the user-supplied postal code and ignoring the other billing data when it's submitted but I certainly could be wrong.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 6 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Good point. I forgot to handle the "processing" state. I think I can do it without this object though, which I felt was an unnecessary abstraction.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 7 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Good point. I'll make something.

File: client/my-sites/checkout/checkout/web-payment-box.jsx [hunk 9]
```diff
@@ -28,523 +26,373 @@ import {
...
+	translate,
+} ) {
+	if ( isStripeLoading ) {
+		return translate( 'Loading…' );
+	}
+	if ( ! canMakePayment ) {
+		return translate( 'This payment method is unavailable' );
+	}
+	if ( ! postalCode || ! countryCode ) {
+		return translate( 'Please specify a country and postal code' );
```
Comment (a8ci18n) [inline client/my-sites/checkout/checkout/web-payment-box.jsx]: Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Please specify a country and postal code.' )` *ES Score: 9*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Correct, but the way this is formatted I think it would be better without the period.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 7 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Done.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 6 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Done.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 3 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Done.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 4 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Done.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 5 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: > This change in behavior is kind of bad because it means we're requiring Apple Pay users to type in lots of extra information

@DavidRothstein I'd like to know more about what you discovered about this since I don't see that I have to enter any information at all during checkout with this option. I just choose a card from the list, and when I entered the card in the first place (using my phone, which AFAIK is the only way to do it), it didn't ask for the billing address at all (it might get it from the credit card processor, I suppose). Maybe there's different flows depending on the card?

> the postal code that the user typed before opening Apple Pay being overwritten (in the data saved server-side) with the one that came from the billing address in Apple Pay

That could be... I haven't checked the server side code yet, but I'm not so sure because on the client we grab the `payerName` and `paymentMethod` and then (in `completePaymentMethodTransaction()`) combine them with the user-selected `postalCode` and `countryCode` to send to the TransactionStore and thence to the server. Was there somewhere specific you saw this happen or suspected it of happening? (There is an unclear object merge right before the call to `completePaymentMethodTransaction` which could in theory overwrite the user supplied values that I'm going to make more explicit, but I checked and the path to the postal data is quite different from our key.)

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 8 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Ah, good find. I had to trace a lot of code to notice that, but I think you're right.

File: client/blocks/credit-card-form/helpers.js (hunk 2 above)
Comment (sirbrillig) [reply client/blocks/credit-card-form/helpers.js]: Done.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 8 above)
Comment (sirbrillig) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Done.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 5 above)
Comment (DavidRothstein) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: >I just choose a card from the list, and when I entered the card in the first place (using my phone, which AFAIK is the only way to do it), it didn't ask for the billing address at all (it might get it from the credit card processor, I suppose). Maybe there's different flows depending on the card?

Hm, I added my card a long time ago, but I'm pretty sure I typed in my billing address by hand at the time.  (I didn't need it for anything, but I seem to remember that the field was empty and I filled it out while poking around the Apple Wallet.)  I wonder what other people experience?

To test it with this pull request, I just deleted the billing address associated with the card in my Apple Wallet and then confirmed that it forced me to type a new one in before I could purchase on WordPress.com.

>>the postal code that the user typed before opening Apple Pay being overwritten (in the data saved server-side) with the one that came from the billing address in Apple Pay
>
>Was there somewhere specific you saw this happen or suspected it of happening?

You can reproduce it with this pull request by typing a ZIP code in the WordPress.com form field that's different from the card's ZIP code in Apple Pay, then purchasing. If you look at the card record server-side (on WordPress.com), it will have the billing ZIP code from Apple Pay, not the one you typed in.

I noticed that if you go into the Stripe dashboard (https://dashboard.stripe.com) and look at the record there, they have the entire billing address from Apple Pay saved also.  So I think what might be happening here is that it's not anything in the JavaScript doing it; rather, Stripe gets the billing address directly from Apple Pay and saves it with the payment method, and then later the WordPress.com code on the server updates its own records with the data from Stripe (overwriting whatever it had for the ZIP code before).

**Comment by sirbrillig** [comment]
I think I've fixed how the card is stored on the server in D32510-code. I still need to figure out the postal code issue, but it might need to be fixed in the same place.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 5 above)
Comment (DavidRothstein) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: I think we could do that if we have to, but it basically has the opposite problem -- everyone will have to type their name in (rather than letting Apple Pay provide it to us automatically).

Maybe a good first step would be to talk with Stripe about why their API behaves this way and figure out f there's a way to get around it...

Comment (yoavf) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: I can confirm the zip code issue is still present - we save the user input country, and the Apple Pay billing address zip in our stored details. I'll take a look.

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 9 above)
Comment (yoavf) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: I've fixed the other occurrence, so we can still end up with just one string versions (the translations should be copied over)

File: client/my-sites/checkout/checkout/web-payment-box.jsx (hunk 5 above)
Comment (yoavf) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: Thinking about it again - we do want to store the proper zip code because it is used as a signal for card verification (especially in the US), and missing it might harm reuse/renewals of the card. I guess we'd want to set the card billing country as well. 

In terms of UX, the UI for these payments is going to change with our expected redesign, so when the steps will be seperate, it will be less of a problem.

Comment (yoavf) [reply client/my-sites/checkout/checkout/web-payment-box.jsx]: We've fixed this in the backend - for now, we'll update the PM on Stripe with the address (zip/country) the user has specifically entered in the form.

**Comment by yoavf** [comment]
With the last remaining review item fixed on the backend, I'm going to merge this and watch closely today make sure Apple Pay is still working as expected.
